v2/devdoc/logging.md

```diff
@@ -0,0 +1,339 @@
+
+- `INFO`
+
+- `VERBOSE`
+
+## Sinks
+
+Multiple sinks shall be supported. Setting up globally which sinks to be used shall be supported (so that they do not have to be specified every time a `c_logging_log` call is made).
+
+Defining the sinks to use shall be done by defining a NULL terminated array, where each entry shall be a pointer to a structure that represents the sink interface.
+
+Example:
+
+```c
+const LOG_SINK* log_sinks[] = { &console_log_sink, &etw_log_sink, NULL };
+```
+
+### ETW sink
+
+An `ETW` sink shall be supported so that events are emitted via TraceLogging.
+
+For the `ETW` sink there should be a posibility of specifying the provider GUID to use.
+
+### console sink
+
+A console sink shall be supported so that events are printed to the console using `printf`.
+
+## API
+
+### logger_log
+
+```c
+void logger_log(LOG_LEVEL log_level, LOG_CONTEXT_HANDLE log_context, const char* file, const char* func, int line_no, const char* format, ...);
+```
+
+`log_context` may be `NULL`. If `log_context` is `NULL`, no context information shall be added to the log line.
+
+`format` shall be a `printf`-like format string.
+
+`...` shall be the variable number of arguments required by `format`.
+
+### LOGGER_LOG
+
+```c
+#define LOGGER_LOG(log_level, log_context, format, ...) \
+    ...
+```
+
+`LOGGER_LOG` is a helper function that automatically adds the file, function and line number information to the log line.
+
+It calls `logger_log` under the hood with the arguments passed by the user while adding the file, function and line number information.
+
+## Contextual logging
+
+For each log line the library shall support specifying a context to be added to the log line.
+
+The context shall allow defining property-value pairs that are added to the event when the event is emitted.
+
+The properties are evaluated at context construction definition time (be that dynamiccally allocated or stack context).
+
+Once the context is built, the properties cannot be changed in the context.
+
+### Local stack context creation
+
+The library shall support defining a context using only stack memory, not requiring any memory allocation:
+
+```c
+#define LOG_CONTEXT_LOCAL_DEFINE(log_context, parent_context, ...) \
+```
+
+where `...` is a list of context fields defined using `LOG_CONTEXT_PROPERTY` or `LOG_CONTEXT_STRING_PROPERTY` entries.
+
+Example:
+
+```c
+    LOG_CONTEXT_LOCAL_DEFINE(log_context, NULL, LOG_CONTEXT_STRING_PROPERTY(property_name, "%s", MU_P_OR_NULL(prop_value)));
+```
+
+Bonus:
+
+An ideal implementation for the stack allocated contexts is also to not do any memory copy until actually needed (until they are used).
+
+That would allow writing code like below while not performing the memory copies until the logging is actually executed.
+
+```c
+void a(LOG_CONTEXT_HANDLE log_context, const char* correlation_id)
+{
+    LOG_CONTEXT_LOCAL_DEFINE(local_context, NULL, LOG_CONTEXT_STRING_PROPERTY(correlation_id, "%s", correlation_id));
```

Comment:
>correlation_id [](http://example.com/codeflow?start=100&length=14)

in this exmaple, can "correlation_id actually be a function name with parameters? And when would that function be evaluated?

Reply:
Updated verbiage about the "Bonus".
We should aim at doing as little work as possible until the log happens, but the evaluation of the property builder macros has to be done at the time of context creation. Otherwise hell breaks loose.